enzyme/tools/enzyme-tblgen/datastructures.cpp

```diff
@@ -0,0 +1,324 @@
+    } else if (val->getName() == "mld") {
+      argTypes.insert(std::make_pair(pos, argType::mldData));
+      argTypes.insert(std::make_pair(pos + 1, argType::mldLD));
+    } else {
+      llvm::errs() << "val->getName: " << val->getName() << "\n";
+      PrintFatalError("Unknown type!");
+    }
+    pos += val->getValueAsInt("nelem");
+  }
+}
+
+void fillArgs(const Record *r, SmallVector<std::string, 6> &args,
+              StringMap<size_t> &argNameToPos) {
+  DagInit *argOps = r->getValueAsDag("PatternToMatch");
+  size_t numArgs = argOps->getNumArgs();
+  args.reserve(numArgs);
+  for (size_t i = 0; i < numArgs; i++) {
+    args.push_back(argOps->getArgNameStr(i).str());
+    argNameToPos.insert(std::pair<std::string, size_t>(args[i], i));
+  }
+  assert(args.size() == numArgs);
+  assert(argNameToPos.size() == numArgs);
+}
+
+void fillArgUserMap(SmallVector<Rule, 3> &rules,
+                    const SmallVector<std::string, 6> &nameVec,
+                    const SmallVector<size_t, 4> &activeArgs,
+                    DenseMap<size_t, DenseSet<size_t>> &argUsers) {
+
+  for (size_t i = 0; i < nameVec.size(); i++) {
+    auto name = nameVec[i];
+    DenseSet<size_t> set{};
+    for (auto rule : llvm::enumerate(rules)) {
+      auto nameMap = rule.value().getArgNameMap();
+      if (nameMap.count(name) == 1) {
+        size_t val = activeArgs[rule.index()];
+        set.insert(val);
+      }
+    }
+    auto newVal = std::make_pair<>(i, set);
+    argUsers.insert(newVal);
+  }
+}
+
+TGPattern::TGPattern(Record &r) {
+  blasName = r.getNameInitAsString();
+
+  args = llvm::SmallVector<std::string, 6>();
+  argNameToPos = StringMap<size_t>{};
+  fillArgs(&r, args, argNameToPos);
+
+  argTypes = DenseMap<size_t, argType>();
+  fillArgTypes(&r, argTypes);
+  if (argTypes.lookup(0) == argType::cblas_layout) {
+    BLASLevel2or3 = true;
+  } else {
+    BLASLevel2or3 = false;
+  }
+
+  posActArgs = SmallVector<size_t, 4>();
+  fillActiveArgSet(&r, posActArgs);
+
+  mutables = DenseSet<size_t>();
+  fillMutableArgSet(&r, mutables);
+
+  // Now create the rules for this pattern
+  {
+    rules = llvm::SmallVector<Rule, 3>{};
+    ListInit *derivOps = r.getValueAsListInit("ArgDerivatives");
+    for (auto derivOp : llvm::enumerate(*derivOps)) {
+      DagInit *derivRule = cast<DagInit>(derivOp.value());
+      size_t actIdx = posActArgs[derivOp.index()];
+      rules.push_back(
+          Rule(derivRule, actIdx, argNameToPos, argTypes, mutables));
+    }
+  }
+
+  argUsers = DenseMap<size_t, DenseSet<size_t>>();
+  fillArgUserMap(rules, args, posActArgs, argUsers);
+}
+SmallVector<size_t, 2> TGPattern::getRelatedLengthArgs(size_t arg) {
+  if (!BLASLevel2or3) {
+    assert(argTypes.lookup(arg) == argType::vincData);
+    assert(argTypes.lookup(0) == argType::len);
+    return {0};
+  }
+  assert(argTypes.lookup(arg) == argType::vincData ||
+         argTypes.lookup(arg) == argType::mldData);
+  // This is terribly wrong because it will burn
```

Comment:
Is this still true?

Reply:
I think so. The issue is that some (A,B), but not all matrices (C) are getting transposed, 
so either I hardcode more combinations and add some logic, or I just extend the .td
file to link trans, dim1 and dim2 from the input args to the corresponding matrices.
Second would be nicer, but too much work for this PR.
Shall I add an assertion for the meantime?

Reply:
E.g. I have something like the following in mind:

```
def gemm : CallBlasPattern<(Op $layout, $transa, $transb, $m, $n, $k, $alpha, $A, $lda, $B, $ldb, $beta, $C, $ldc),
                  ["C"],
                  [cblas_layout, trans, trans, len, len, len, fp, mld, mld, fp, mld],
                  [
                  /* alpha */ (inactive),
                  /* A     */ (b<"gemm"> $layout, $transa, transpose<"transb">, $m, $k, $n, $alpha, adj<"C">, $ldc, $B, (ld $B, $transb, $ldb, $k, $n), $beta, adj<"A">, $lda),
                  /* B     */ (b<"gemm"> $layout, transpose<"transa">, $transb, $k, $n, $m, $alpha, $A, (ld $A, $transa, $lda, $m, $k), adj<"C">, $ldc, $beta, adj<"B">, $ldb),
                  /* beta  */ (inactive),
                  /* C     */ (b<"lascl"> $layout, Char<"G">, ConstantInt<"0">, ConstantInt<"0">, Constant<"1.0">, $beta, $m, $n, adj<"C">, $ldc, ConstantInt<"0">)
                  ]
                  >;
```
becomes 
```
def gemm : CallBlasPattern<(Op $layout, $transa, $transb, $m, $n, $k, $alpha, $A, $lda, $B, $ldb, $beta, $C, $ldc),
                  ["C"],
                  [cblas_layout, trans, trans, len, len, len, fp, mld<$A, $transa, $m, $n>, mld<$B, $transb, $n, $k>, fp, mld<$C, $m, $k>],
```
I have to think a bit more about it though, since in some cases the value of trans also affects the number of elements in a vector like x, because it might either be m or n.
That's why I want to do that outside of this already to large PR.

Reply:
Yeah an assertion would be fine. I'm right now a big fan of always correct if it works, and perhaps over eager errors otherwise.